### PR TITLE
feat(wandb): add logging to W&B

### PR DESCRIPTION
**Pull request by borisdayma**
Hi,

I've been using hugging-face for a while and I've been loving it. I think it could become a central reference for NLP problems. I would like to contribute in solving a few limitations I observed:
* some models don't have clear details on how they have been trained (hyper-parameters) and for which task, making it hard to reproduce their results (and even ensure those results were actually reached) -> ideally we would track & log their training process as well as config parameters used
* some new models keep on being requested/proposed without seeing their actual efficiency -> ideally we would be able to compare them to baselines
* I've seen issues where "better" models are offered to replace previous ones but there is no way to prove they are actually better -> maybe we can add a metrics table (linked to actual runs) to the model card
* it is hard to keep track of models and datasets
* overall I'd like to bring more traceability and transparency between users

This is why I propose to use [Weights & Biases](https://docs.wandb.com/) which is free for open source and will help solve those issues. Also it is already part of Pytorch-Lightning so it will be easy to integrate it with those scripts.

This PR brings :

* arg `--log_on_wandb` will save all config parameters and training/evaluation metrics to W&B
* arg `--upload_model` will upload trained model

See example run on bert-base-cased  -> [W&B run](https://app.wandb.ai/borisd13/transformers-examples/runs/2hkb07jl)

image

If you navigate around, you will see that config parameters, metrics, gradients, computer resources, trained model, actual code (including git repo & command line used) and stdout have all been uploaded for full traceability and reproducibility.

I can also easily compare bert-base-cased and bert-base-uncased from my [project page](https://app.wandb.ai/borisd13/transformers-examples).

image

This makes it easy to quickly see if a new model is actually interesting.

Finally, this integration let us create [cool reports](https://app.wandb.ai/stacey/keras_finetune/reports/Curriculum-Learning-in-Nature--Vmlldzo1MjcxNw) where graphs directly pool data and are interactive.

If you like it, then I'd love to add other features in later PR's:

* in addition to metrics, track model size & inference speed
* use artifacts (their version control system) to track models & datasets -> central repository that will let us see easily that runs are based on same dataset or same pre-trained model + see all trained tasks associated to one pre-trained model
* log prediction samples (based on the task)
* create a central repo on W&B (entity: hugging-face) to have access to other's people runs
* add the same changes to remaining scripts (should be straight forward as it's just a few lines)
* integrate model card issue template so that we always add a link to a run as well as performance metrics
* add [sweeps](https://docs.wandb.com/sweeps) -> make it easy to optimize hyper-parameters
* we could add a hook to run automatically on all tasks any new model added (and link the run to the model cards)

Please let me know your thoughts and feel free to contact me or Weights & Biases directly to discuss this integration.

### CODE REVIEW

File: examples/run_glue.py [hunk 1]
```diff
@@ -505,6 +515,16 @@ def main():
...
 
+    # Start a wandb run
+    if args.log_on_wandb:
+        if not _WANDB_AVAILABLE:
+            logger.error("To use Weights & Biases logging, import wandb.")
+        else:
+            wandb.init(config={**vars(args), 'script': os.path.basename(__file__), 'config':config.to_dict(),
+                               'tokenizer_config':tokenizer.init_kwargs})
+        # keep track of model topology and gradients
+        wandb.watch(model)
```
Comment (vanpelt) [inline examples/run_glue.py]: Don't call watch if not `_WANDB_AVAILABLE`

File: examples/run_glue.py [hunk 2]
```diff
@@ -529,6 +549,10 @@ def main():
         # Good practice: save your training arguments together with the trained model
         torch.save(args, os.path.join(args.output_dir, "training_args.bin"))
 
+        # Upload trained model and relevant files (no checkpoints)
+        if args.log_on_wandb and args.upload_model and _WANDB_AVAILABLE:
+            wandb.save(os.path.join(args.output_dir, "*.*"))
```
Comment (vanpelt) [inline examples/run_glue.py:554]: Are there any other files besides the model and training_args in the output dir?

File: examples/run_glue.py [hunk 3]
```diff
@@ -227,6 +232,11 @@ def train(args, train_dataset, model, tokenizer):
 
                     for key, value in logs.items():
                         tb_writer.add_scalar(key, value, global_step)
+
+                    if args.log_on_wandb and _WANDB_AVAILABLE:
+                        epoch = global_step / (len(train_dataloader) // args.gradient_accumulation_steps)
+                        wandb.log({**logs, "epoch": epoch})
```
Comment (vanpelt) [inline examples/run_glue.py:238]: We could add `step=global_step` here for good measure.

**Review by vanpelt — APPROVED** [review]
Looks good.  Do you envision adding this to all example models?  A potentially broader integration would be to hook into something like `from_pretrained` and call init from there.

**Comment by borisdayma** [comment]
> Looks good. Do you envision adding this to all example models? A potentially broader integration would be to hook into something like `from_pretrained` and call init from there.

I didn't think about this one. There are a lot of functions `from_pretrained` though (torch, tf, type of model…) and I also need the tokenizer details so I thought the best place would be to do all the scripts (example models).

There is currently a [rewrite of the API](https://github.com/huggingface/transformers/pull/3800) so there may be a more central place to do it. Now it's mainly for tensorboard and a bit limiting (only log metrics) for what we can do with wandb (log models, watch gradients, log prediction samples…) but we can find ways to improve it.

File: examples/run_glue.py (hunk 1 above)
Comment (borisdayma) [reply examples/run_glue.py]: Good catch

File: examples/run_glue.py (hunk 2 above)
Comment (borisdayma) [reply examples/run_glue.py:554]: In theory no (except for a small sample output file) unless using `--overwrite_output_dir` but the more robust way will be to get file names from the save functions (will require to update `save_pretrained` function to return file paths used for model and config files)

File: examples/run_glue.py (hunk 3 above)
Comment (borisdayma) [reply examples/run_glue.py:238]: The issue I've been having with `step` is that it does not commit when using it until next step so I need to add a `wandb.log({})` to update my last step but I agree it would be nice to see it.
